File: packages/kits/GeniusLocusKit/rust/tests/fact_extraction_duty_tests.rs
use std::sync::Arc;

use fact_extraction_kit::contract::{
    FactAssertionKind, FactCandidate, FactExtractionError, FactExtractionRequest,
    FactExtractionResponse, FactExtractor, FactExtractorKind, FactExtractorModelSpec,
};
use genius_locus_kit::EstateCoordinator;
use locus_kit::drawer_operational::CaptureChannel;
use locus_kit::drawer_store::DrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::estate_types::{LatticeAnchor, OwnerCredentials};
use locus_kit::frames::CaptureFrame;
use locus_kit::kg_fact::KGFactOrigin;

const NOW: i64 = 1_700_000_000;
const SOURCE: &str = "Jack's birthday is June 20th.";

struct FakeExtractor {
    spec: FactExtractorModelSpec,
    empty: bool,
    grounded: bool,
}

impl FactExtractor for FakeExtractor {
    fn spec(&self) -> &FactExtractorModelSpec {
        &self.spec
    }

    fn extract(
        &self,
        request: &FactExtractionRequest,
    ) -> Result<FactExtractionResponse, FactExtractionError> {
        Ok(FactExtractionResponse {
            source_digest: request.source_digest.clone(),
            provider_id: self.spec.provider_id.clone(),
            model_id: self.spec.model_id.clone(),
            model_version: self.spec.model_version.clone(),
            schema_version: self.spec.schema_version.clone(),
            candidates: if self.empty {
                vec![]
            } else {
                vec![FactCandidate {
                    subject: "Jack".into(),
                    predicate: "birthday".into(),
                    object: if self.grounded {
                        "June 20th"
                    } else {
                        "July 4th"
                    }
                    .into(),
                    evidence_quote: if self.grounded {
                        SOURCE
                    } else {
                        "Jack's birthday is July 4th."
                    }
                    .into(),
                    confidence: 0.97,
                    assertion_kind: FactAssertionKind::Asserted,
                    search_aliases: vec!["when is Jack's birthday".into()],
                }]
            },
        })
    }
}

fn spec() -> FactExtractorModelSpec {
    FactExtractorModelSpec {
        provider_id: "test-provider".into(),
        model_id: "nuextract-test".into(),
        model_version: "q8".into(),
        schema_version: "kgfact-extraction-v1".into(),
        extractor_kind: FactExtractorKind::SpecializedModel,
        maximum_input_characters: 16_384,
        maximum_facts_per_source: 8,
    }
}

fn open() -> (
    EstateCoordinator,
    genius_locus_kit::EstateHandle,
    Arc<InMemoryDrawerStore>,
) {
    let mut coordinator = EstateCoordinator::new();
    let store = Arc::new(InMemoryDrawerStore::new(NOW, None).expect("store"));
    let handle = coordinator
        .open(
            Arc::clone(&store) as Arc<dyn DrawerStore>,
            OwnerCredentials::new("fact-duty"),
            0,
            100,
        )
        .expect("open");
    (coordinator, handle, store)
}

fn capture(
    coordinator: &EstateCoordinator,
    handle: &genius_locus_kit::EstateHandle,
    body: &str,
) -> String {
    coordinator
        .capture(
            handle,
            CaptureFrame::new(
                body,
                CaptureChannel::Typed,
                "facts",
                LatticeAnchor::udc("000"),
                "test",
                "test-v1",
            ),
            NOW,
        )
        .expect("capture")
        .id
}

#[test]
fn grounded_fact_files_with_provenance_and_settles_debt() {
    let (mut coordinator, handle, store) = open();
    let drawer_id = capture(&coordinator, &handle, SOURCE);
    coordinator
        .activate_fact_extractor(
            Arc::new(FakeExtractor {
                spec: spec(),
                empty: false,
                grounded: true,
            }),
            "nuextract-b1-q8-v1",
            &handle,
        )
        .expect("activate");
    let report = coordinator
        .run_fact_extraction_batch(&handle, 16, NOW)
        .expect("batch");
    assert_eq!(report.completed_sources, 1);
    assert_eq!(report.facts_filed, 1);
    assert_eq!(report.failed_sources, 0);

    let facts = store.all_kg_facts().expect("facts");
    let fact = &facts[0];
    assert_eq!(fact.source_drawer_id, drawer_id);
    assert_eq!(fact.evidence_quote, SOURCE);
    assert_eq!(fact.evidence_start, 0);
    assert_eq!(fact.evidence_end, SOURCE.chars().count() as i64);
    assert_eq!(fact.extractor_model_id, "nuextract-test");
    assert_eq!(fact.search_projection_version, "kgfact-search-v1");
    assert!(fact.search_projection.contains("when is Jack's birthday"));
    assert!(store
        .get_drawer(&drawer_id)
        .unwrap()
        .unwrap()
        .are_facts_extracted());

    let replay = coordinator
        .run_fact_extraction_batch(&handle, 16, NOW)
        .expect("replay");
    assert_eq!(replay.completed_sources, 0);
    assert_eq!(replay.facts_filed, 0);
}

#[test]
fn empty_response_is_a_valid_zero_fact_completion() {
    let (mut coordinator, handle, store) = open();
    let drawer_id = capture(&coordinator, &handle, "A friendly hello.");
    coordinator
        .activate_fact_extractor(
            Arc::new(FakeExtractor {
                spec: spec(),
                empty: true,
                grounded: true,
            }),
            "nuextract-b1-q8-v1",
            &handle,
        )
        .expect("activate");
    let report = coordinator
        .run_fact_extraction_batch(&handle, 16, NOW)
        .expect("batch");
    assert_eq!(report.completed_sources, 1);
    assert_eq!(report.facts_filed, 0);
    assert!(store.all_kg_facts().unwrap().is_empty());
    assert!(store
        .get_drawer(&drawer_id)
        .unwrap()
        .unwrap()
        .are_facts_extracted());
}

#[test]
fn wholly_ungrounded_output_stays_debt() {
    let (mut coordinator, handle, store) = open();
    let drawer_id = capture(&coordinator, &handle, SOURCE);
    coordinator
        .activate_fact_extractor(
            Arc::new(FakeExtractor {
                spec: spec(),
                empty: false,
                grounded: false,
            }),
            "nuextract-b1-q8-v1",
            &handle,
        )
        .unwrap();
    let result = coordinator
        .run_fact_extraction_batch(&handle, 16, NOW)
        .unwrap();
    assert_eq!(result.failed_sources, 1);
    assert_eq!(result.facts_filed, 0);
    assert!(!store
        .get_drawer(&drawer_id)
        .unwrap()
        .unwrap()
        .are_facts_extracted());
    assert!(store.all_kg_facts().unwrap().is_empty());
}

#[test]
fn recipe_replacement_retires_machine_fact_but_preserves_manual_fact() {
    let (mut coordinator, handle, store) = open();
    let drawer_id = capture(&coordinator, &handle, SOURCE);
    coordinator
        .activate_fact_extractor(
            Arc::new(FakeExtractor {
                spec: spec(),
                empty: false,
                grounded: true,
            }),
            "nuextract-b1-q8-v1",
            &handle,
        )
        .unwrap();
    coordinator
        .run_fact_extraction_batch(&handle, 16, NOW)
        .unwrap();
    let old_machine_id = store.all_kg_facts().unwrap()[0].id.clone();
    let manual = coordinator
        .add_kg_fact_with_id_and_origin(
            &handle,
            "manual-jack-birthday-note",
            "Jack",
            "birthday-note",
            "confirmed by Bob",
            &drawer_id,
            &KGFactOrigin {
                added_by: "human".into(),
                ..KGFactOrigin::default()
            },
            NOW,
        )
        .unwrap();

    let mut replacement_spec = spec();
    replacement_spec.model_id = "nuextract-replacement".into();
    replacement_spec.schema_version = "kgfact-extraction-v2".into();
    assert_eq!(
        coordinator
            .activate_fact_extractor(
                Arc::new(FakeExtractor {
                    spec: replacement_spec,
                    empty: false,
                    grounded: true,
                }),
                "nuextract-replacement-v2",
                &handle,
            )
            .unwrap(),
        1
    );
    let report = coordinator
        .run_fact_extraction_batch(&handle, 16, NOW)
        .unwrap();
    assert_eq!(report.completed_sources, 1);
    assert_eq!(report.facts_filed, 1);

    let active = store.all_kg_facts().unwrap();
    assert!(active.iter().any(|fact| fact.id == manual.id));
    assert!(!active.iter().any(|fact| fact.id == old_machine_id));
    assert!(active
        .iter()
        .any(|fact| fact.extractor_model_id == "nuextract-replacement"));
    let history = store.all_kg_facts_including_retired().unwrap();
    assert_eq!(history.len(), 3);
    assert!(history.iter().any(|fact| fact.id == old_machine_id));
}

File: packages/kits/GeniusLocusKit/Tests/GeniusLocusKitTests/FactExtractionDutyTests.swift
import FactExtractionKit
import FactExtractionKitProviders
import Foundation
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
import Testing
@testable import GeniusLocusKit

@Suite("Source-grounded fact extraction duty", .serialized)
struct FactExtractionDutyTests {
    private let now = Date(timeIntervalSince1970: 1_700_000_000)

    private func openEstate(
        owner: String
    ) async throws -> (GeniusLocusKit, EstateHandle) {
        let kit = GeniusLocusKit()
        let storage = InMemoryStorage(configuration: EstateConfiguration(
            estateID: UUID(), backend: .inMemory))
        let credentials = OwnerCredentials(ownerIdentifier: owner)
        _ = try await LocusKit.Estate.create(storage: storage, owner: credentials)
        let handle = try await kit.open(
            storage: storage, owner: credentials,
            identityKeyStore: InMemoryEstateIdentityKeyStore())
        return (kit, handle)
    }

    private func spec(_ kind: FactExtractorKind = .specializedModel) -> FactExtractorModelSpec {
        FactExtractorModelSpec(
            providerID: "test-provider", modelID: "nuextract-test",
            modelVersion: "q8", schemaVersion: "kgfact-extraction-v1",
            extractorKind: kind, maximumInputCharacters: 16_384,
            maximumFactsPerSource: 8)
    }

    private func capture(
        _ kit: GeniusLocusKit, _ handle: EstateHandle, content: String
    ) async throws -> Drawer {
        try await kit.capture(handle, CaptureFrame(
            content: content, channel: .typed, room: "facts",
            latticeAnchor: LatticeAnchor(udcCode: "000"), addedBy: "test",
            embeddingModelID: "test-v1", eventTime: now))
    }

    @Test("files a grounded KGFact with provenance and settles debt")
    func filesGroundedFact() async throws {
        let (kit, handle) = try await openEstate(owner: "fact-grounded")
        let source = "Jack's birthday is June 20th."
        let drawer = try await capture(kit, handle, content: source)
        let model = spec()
        let extractor = ClosureFactExtractor(spec: model) { request in
            #expect(request.sourceText.contains("Jack"))
            #expect(!request.eligibleSourceSpans.isEmpty)
            return FactExtractionResponse(
                sourceDigest: request.sourceDigest,
                providerID: model.providerID, modelID: model.modelID,
                modelVersion: model.modelVersion, schemaVersion: model.schemaVersion,
                candidates: [FactCandidate(
                    subject: "Jack", predicate: "birthday", object: "June 20th",
                    evidenceQuote: source, confidence: 0.97,
                    searchAliases: ["Jack birthday", "when is Jack's birthday"])])
        }
        _ = try await kit.activateFactExtractor(
            extractor, recipeID: "nuextract-b1-q8-v1", for: handle)
        let report = try await kit.runFactExtractionBatch(handle, now: now)
        #expect(report == FactExtractionBatchResult(
            completedSources: 1, factsFiled: 1, candidatesRejected: 0,
            skippedSources: 0, failedSources: 0))

        let estate = try await kit.estate(for: handle)
        let fact = try #require(try await estate.allKGFacts().first)
        #expect(UUID(uuidString: fact.id) != nil)
        #expect(fact.sourceDrawerID == drawer.id)
        #expect(fact.evidenceQuote == source)
        #expect(fact.evidenceStart == 0 && fact.evidenceEnd == source.unicodeScalars.count)
        #expect(fact.sourceDigest.count == 64)
        #expect(fact.extractorModelID == model.modelID)
        #expect(fact.searchProjection.contains("when is Jack's birthday"))
        #expect(fact.searchProjectionVersion == FactSearchProjection.version)
        #expect(fact.extractorClass == .specializedModel)
        #expect(fact.confidenceBand == .certain)
        #expect(try await estate.getDrawers(ids: [drawer.id]).first?.areFactsExtracted == true)

        let replay = try await kit.runFactExtractionBatch(handle, now: now)
        #expect(replay.completedSources == 0 && replay.factsFiled == 0)
    }

    @Test("empty model result is settled, wholly ungrounded output stays debt")
    func zeroAndUngroundedOutcomes() async throws {
        let (zeroKit, zeroHandle) = try await openEstate(owner: "fact-zero")
        let zeroDrawer = try await capture(zeroKit, zeroHandle, content: "A friendly hello.")
        let model = spec(.foundationModel)
        let empty = ClosureFactExtractor(spec: model) { request in
            FactExtractionResponse(
                sourceDigest: request.sourceDigest,
                providerID: model.providerID, modelID: model.modelID,
                modelVersion: model.modelVersion, schemaVersion: model.schemaVersion,
                candidates: [])
        }
        _ = try await zeroKit.activateFactExtractor(
            empty, recipeID: "apple-system-v1", for: zeroHandle)
        let zero = try await zeroKit.runFactExtractionBatch(zeroHandle, now: now)
        #expect(zero.completedSources == 1 && zero.factsFiled == 0)
        let zeroEstate = try await zeroKit.estate(for: zeroHandle)
        #expect(try await zeroEstate.getDrawers(ids: [zeroDrawer.id]).first?.areFactsExtracted == true)
        #expect(try await zeroEstate.allKGFacts().isEmpty)

        let (badKit, badHandle) = try await openEstate(owner: "fact-bad")
        let badDrawer = try await capture(
            badKit, badHandle, content: "Jack's birthday is June 20th.")
        let bad = ClosureFactExtractor(spec: model) { request in
            FactExtractionResponse(
                sourceDigest: request.sourceDigest,
                providerID: model.providerID, modelID: model.modelID,
                modelVersion: model.modelVersion, schemaVersion: model.schemaVersion,
                candidates: [FactCandidate(
                    subject: "Jack", predicate: "birthday", object: "July 4th",
                    evidenceQuote: "Jack's birthday is July 4th.", confidence: 0.99)])
        }
        _ = try await badKit.activateFactExtractor(
            bad, recipeID: "apple-system-v1", for: badHandle)
        let rejected = try await badKit.runFactExtractionBatch(badHandle, now: now)
        #expect(rejected.failedSources == 1 && rejected.candidatesRejected == 1)
        let badEstate = try await badKit.estate(for: badHandle)
        #expect(try await badEstate.getDrawers(ids: [badDrawer.id]).first?.areFactsExtracted == false)
        #expect(try await badEstate.allKGFacts().isEmpty)
    }

    @Test("recipe replacement retires prior machine facts but preserves manual facts")
    func recipeReplacementPreservesManualFacts() async throws {
        let (kit, handle) = try await openEstate(owner: "fact-recipe-replacement")
        let source = "Jack's birthday is June 20th."
        let drawer = try await capture(kit, handle, content: source)
        let firstModel = spec()
        let first = ClosureFactExtractor(spec: firstModel) { request in
            FactExtractionResponse(
                sourceDigest: request.sourceDigest,
                providerID: firstModel.providerID, modelID: firstModel.modelID,
                modelVersion: firstModel.modelVersion,
                schemaVersion: firstModel.schemaVersion,
                candidates: [FactCandidate(
                    subject: "Jack", predicate: "birthday", object: "June 20th",
                    evidenceQuote: source, confidence: 0.97)])
        }
        _ = try await kit.activateFactExtractor(
            first, recipeID: "nuextract-b1-q8-v1", for: handle)
        _ = try await kit.runFactExtractionBatch(handle, now: now)
        let estate = try await kit.estate(for: handle)
        let oldMachineID = try #require(try await estate.allKGFacts().first).id

        let manual = try await kit.captureKGFact(
            handle, id: "manual-jack-birthday-note", subject: "Jack",
            predicate: "birthday-note", object: "confirmed by Bob",
            sourceDrawerID: drawer.id, addedBy: "human", now: now)

        let replacementModel = FactExtractorModelSpec(
            providerID: "test-provider", modelID: "nuextract-replacement",
            modelVersion: "q8", schemaVersion: "kgfact-extraction-v2",
            extractorKind: .specializedModel, maximumInputCharacters: 16_384,
            maximumFactsPerSource: 8)
        let replacement = ClosureFactExtractor(spec: replacementModel) { request in
            FactExtractionResponse(
                sourceDigest: request.sourceDigest,
                providerID: replacementModel.providerID,
                modelID: replacementModel.modelID,
                modelVersion: replacementModel.modelVersion,
                schemaVersion: replacementModel.schemaVersion,
                candidates: [FactCandidate(
                    subject: "Jack", predicate: "birthday", object: "June 20th",
                    evidenceQuote: source, confidence: 0.97)])
        }
        #expect(try await kit.activateFactExtractor(
            replacement, recipeID: "nuextract-replacement-v2", for: handle) == 1)
        let report = try await kit.runFactExtractionBatch(handle, now: now)
        #expect(report.completedSources == 1 && report.factsFiled == 1)

        let active = try await estate.allKGFacts()
        #expect(active.contains(where: { $0.id == manual.id }))
        #expect(!active.contains(where: { $0.id == oldMachineID }))
        #expect(active.contains(where: {
            $0.extractorModelID == replacementModel.modelID
        }))
        let history = try await estate.allKGFactsIncludingRetired()
        #expect(history.count == 3)
        #expect(history.contains(where: { $0.id == oldMachineID }))
    }

    @Test("source-exact chunking reaches a fact beyond the first model window")
    func extractsTailFactFromOriginalBody() async throws {
        let (kit, handle) = try await openEstate(owner: "fact-original-body")
        let factText = "Jack's birthday is June 20th."
        let source = String(repeating: "Background material. ", count: 40) + factText
        let drawer = try await capture(kit, handle, content: source)
        let model = FactExtractorModelSpec(
            providerID: "test-provider", modelID: "nuextract-test",
            modelVersion: "q8", schemaVersion: "kgfact-extraction-v1",
            extractorKind: .specializedModel, maximumInputCharacters: 700,
            maximumFactsPerSource: 8)
        let extractor = ClosureFactExtractor(spec: model) { request in
            let candidates = request.sourceText.contains(factText) ? [FactCandidate(
                subject: "Jack", predicate: "birthday", object: "June 20th",
                evidenceQuote: factText, confidence: 0.97)] : []
            return FactExtractionResponse(
                sourceDigest: request.sourceDigest,
                providerID: model.providerID, modelID: model.modelID,
                modelVersion: model.modelVersion, schemaVersion: model.schemaVersion,
                candidates: candidates)
        }
        _ = try await kit.activateFactExtractor(
            extractor, recipeID: "nuextract-original-body-v1", for: handle)

        let report = try await kit.runFactExtractionBatch(handle, now: now)
        #expect(report.completedSources == 1)
        #expect(report.factsFiled == 1)
        let fact = try #require(try await kit.estate(for: handle).allKGFacts().first)
        #expect(fact.sourceDrawerID == drawer.id)
        #expect(fact.evidenceQuote == factText)
    }
}
